enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/PropertyExistenceEnforcer.java

```diff
@@ -42,191 +49,406 @@
+                    }
+                } );
+            }
+        }
+        if ( nodes.isEmpty() && relationships.isEmpty() )
+        {
+            return NO_CONSTRAINTS;
+        }
+        return new PropertyExistenceEnforcer( nodes, relationships );
+    };
 
-    private void validateNode( long nodeId ) throws NodePropertyExistenceException
+    private class Decorator extends TxStateVisitor.Delegator
     {
-        if ( labelExistenceConstraints.isEmpty() )
+        private final ReadableTransactionState txState;
+        private final StoreReadLayer storeLayer;
+        private final PrimitiveIntSet propertyKeyIds = Primitive.intSet();
+        private StorageStatement storageStatement;
+
+        Decorator( TxStateVisitor next, ReadableTransactionState txState, StoreReadLayer storeLayer )
+        {
+            super( next );
+            this.txState = txState;
+            this.storeLayer = storeLayer;
+        }
+
+        @Override
+        public void visitNodePropertyChanges(
+                long id, Iterator<StorageProperty> added, Iterator<StorageProperty> changed,
+                Iterator<Integer> removed ) throws ConstraintValidationException
+        {
+            validateNode( id );
+            super.visitNodePropertyChanges( id, added, changed, removed );
+        }
+
+        @Override
+        public void visitNodeLabelChanges( long id, Set<Integer> added, Set<Integer> removed )
+                throws ConstraintValidationException
+        {
+            validateNode( id );
+            super.visitNodeLabelChanges( id, added, removed );
+        }
+
+        @Override
+        public void visitCreatedRelationship( long id, int type, long startNode, long endNode )
+                throws ConstraintValidationException
+        {
+            validateRelationship( id );
+            super.visitCreatedRelationship( id, type, startNode, endNode );
+        }
+
+        @Override
+        public void visitRelPropertyChanges(
+                long id, Iterator<StorageProperty> added, Iterator<StorageProperty> changed,
+                Iterator<Integer> removed ) throws ConstraintValidationException
         {
-            return;
+            validateRelationship( id );
+            super.visitRelPropertyChanges( id, added, changed, removed );
         }
 
-        try ( Cursor<NodeItem> node = nodeCursor( nodeId ) )
+        @Override
+        public void close()
         {
-            if ( node.next() )
+            super.close();
+            if ( storageStatement != null )
             {
-                PrimitiveIntSet labelIds = node.get().labels();
+                storageStatement.close();
+            }
+        }
 
-                propertyKeyIds.clear();
-                try ( Cursor<PropertyItem> properties = properties( node.get() ) )
+        private void validateNode( long nodeId ) throws NodePropertyExistenceException
+        {
+            if ( mandatoryNodePropertiesByLabel.isEmpty() )
```

Comment:
i.e. to move the definition of the `Decorator` class to the bottom of the `PropertyExistenceEnforcer` class (below all the methods)?